### ISSUE TITLE
Easier finger photo on Valve Index Knuckles controllers

### ISSUE THREAD

**Issue by uyjulian** [Yellow-Dog-Man/Resonite-Issues#452]
### Is your feature request related to a problem? Please describe.

On Valve Index Knuckles controllers, it is difficult to fit the conditions for taking the finger photo on the first time, or unintentionally taking a finger photo.

Even if the 3 fingers are touching the grip area, the fingers may not appear all down in-game, in which case turning off and turning back on the controller may allow you to try it again.

It is much easier to do this on Oculus Touch controllers since there is only a single grip button which controls the position of those three fingers in-game.

### Describe the solution you'd like

Instead of reading the skeleton from the steam vr runtime, read the controller data directly.

E.g. if a certain percentage like 75 percent of the grip surface area is touched, then consider it as 3 fingers down for the purpose of the finger photo conditions.
If the index finger is touching the trigger (instead of just hovering over it), consider it as taking the finger photo.

### Describe alternatives you've considered

Replacing the skeleton might be an alternative choice

### Additional Context

n/a

**Comment by Beethoven-n** [comment]
one thing i found that helps is using your lasers to align for the pose you need to do

**Comment by FlameSoulis** [comment]
> one thing i found that helps is using your lasers to align for the pose you need to do

Doesn't dismiss how much of a pain it is to do on an Index. Plus, some avatars have misleading directions for the fingers/laser.

**Comment by Beethoven-n** [comment]
yeah you're right. it should be easier

**Comment by troyBORG** [comment]
Yeah there has been times I have to do the finger trill to reprogram the controllers to get my fingers all down.  Or sometimes because the touch sensor on my left stick is dead my thumb will be down so I don't get the photo until I touch the stick to get it to reset its position.  

I almost would like a thing where you can say "this is the gesture I want to use".

Like how on your phone you have to train your fingerprint unlock,  have it you do the gesture you want multiple times so it learns how you pose your hands for the finger photo to trigger.

**Comment by Frooxius** [comment]
This is a bit tricky, because the gesture is a bit of a sliding scale. Making it easier to trigger also often increases the accidental takes - in fact, it's been made harder to trigger, because people would do it accidentally too many times.

The gesture is based purely on the skeleton though, which is intentional - it makes it work both with controllers and with finger tracking solutions. We are unlikely to move away from this design and implement different solutions for different controllers, because this would require us to design several differing solutions, instead of one unified one.

However the actual logic of the gesture should be reworked a bit, which should help improve how it behaves overall. There's a lot of "wrong" ways to trigger it, which shouldn't be possible, so there's something that is messed up with the logic on triggering the gesture, but I haven't had time to really look into this.

**Comment by Frooxius** [comment]
That would require moving away from the design as well and make this even more complex task.

Instead of a system that's fine tuned for this gesture, we would need to implement a robust system to define custom gestures, which is at least order of magnitude more complex engineering-wise.

**Comment by TisFoolish** [comment]
I've been lead to believe that the finger photo system has an internal set of rotations for the hands and fingers that it compares against and if the the hands and fingers are "close enough" it enables the finger photo, however it seems that may not be the case? It'd be good to know how it actually determines when to turn on and off, it'd at least help in figuring out why some avatars have *incredibly* hard to trigger finger photo.

**Comment by troyBORG** [comment]
Yeah it's mostly positioning on your actual hands.
Like I can use any model.  Doing finger photos with the NeoFur models is possible.

What controllers are you using, I can actually trigger it even easier now on the quest pro controllers vs the index controllers.

**Comment by shadowpanther** [comment]
From my experience:
* Oculus controllers via Oculus runtime (no SteamVR) - good repeatable pose, because finger poses are quite fixed (thumb is either up or down, for example)
* Knuckles on SteamVR - a bit tricky, because finger poses are a continuum, and the controller often recalibrates the touch values/weights
* Quest controllers via SteamLink on SteamVR - can't do the gesture, because for fixed finger poses SteamVR adds a curl depending on hand orientation

**Comment by FlameSoulis** [comment]
>     * Quest controllers via SteamLink on SteamVR - can't do the gesture, because for fixed finger poses SteamVR adds a curl depending on hand orientation

I can confirm this with the Quest Pro via SteamLink. In fact, when I was having trouble, the people in the session suggested turning the right hand by 180 degrees, as that was "the solution to the problem."

The community knows this feature is very problematic and even frustrating to use due to its fickle nature, which is a shame because it is a neat idea.
